src/gateway/sharding/shard_queue.rs

```diff
@@ -0,0 +1,60 @@
+use std::collections::VecDeque;
+use std::num::NonZeroU16;
+
+use super::ShardId;
+use crate::internal::prelude::*;
+
+/// A queue of [`ShardId`]s that is split up into multiple buckets according to the value of
+/// [`max_concurrency`](crate::model::gateway::SessionStartLimit::max_concurrency).
+#[must_use]
+pub struct ShardQueue {
+    buckets: FixedArray<VecDeque<ShardId>, u16>,
+    concurrent: bool,
+}
+
+impl ShardQueue {
+    pub fn new(max_concurrency: NonZeroU16) -> Self {
+        let buckets = vec![VecDeque::new(); max_concurrency.get() as usize].into_boxed_slice();
+        let buckets = FixedArray::try_from(buckets).expect("should fit without truncation");
+
+        Self {
+            buckets,
+            concurrent: false,
+        }
+    }
+
+    fn calculate_bucket(&self, shard_id: ShardId) -> u16 {
+        shard_id.0 % self.buckets.len()
+    }
+
+    /// Calculates the corresponding bucket for the given `ShardId` and **appends** to it.
+    pub fn push_back(&mut self, shard_id: ShardId) {
+        let bucket = self.calculate_bucket(shard_id);
+        self.buckets[bucket].push_back(shard_id);
+    }
+
+    /// Calculates the corresponding bucket for the given `ShardId` and **prepends** to it.
+    pub fn push_front(&mut self, shard_id: ShardId) {
+        let bucket = self.calculate_bucket(shard_id);
+        self.buckets[bucket].push_front(shard_id);
+    }
+
+    /// Pops a `ShardId` from every bucket containing at least one and returns them all as a `Vec`.
+    ///
+    /// If the queue is in concurrent mode, calling this will only return `Some` if all buckets are
+    /// filled and a maximally-sized batch can be popped, otherwise will return `None`.
+    ///
+    /// If the queue is _not_ in concurrent mode, this will return `Some` as long as at least one
+    /// bucket is filled.
+    pub fn pop_batch(&mut self) -> Option<Vec<ShardId>> {
+        if self.concurrent && !self.buckets.iter().all(|b| !b.is_empty()) {
+            None
+        } else {
+            Some(self.buckets.iter_mut().filter_map(VecDeque::pop_front).collect())
+        }
```

Comment:
Performance nit: This currently iterates twice when concurrent is false, when it could just do the one iteration and check if the Vec is empty.

Reply:
How would I do this without actually popping the buckets? If I'm iterating only once, I can't know whether or not to pop from a given bucket because a later bucket might not be filled yet.

Reply:
I'm silly, I misread how the logic flows when concurrent is false.